### PR TITLE
Min stake delegation 0.1

### PR DESCRIPTION
**Pull request by michaelh-laine**
Feature proposal to change the min stake delegation to 0.1 SOL 

Feature gate ID: `3Ynd8eXcJQVc79V75dH1Ut4JR5taoRG7j4oYJKQUpaGd`

Feature to be activated through governance.

### CODE REVIEW

**Comment by michaelh-laine** [comment]
Seems to be failing on a test, not sure why

**Comment by michaelh-laine** [comment]
@brooksprumo sorry to tag, can you take a look at this PR? I've added tests to account for all scenarios (your feature activated, my feature activated, both activated). The CI is failing on an unrelated test not relating to rpc pubsub?

File: programs/stake/src/stake_instruction.rs [hunk 1]
```diff
@@ -529,10 +529,25 @@ mod tests {
...
+        feature_set.deactivate(&feature_set::stake_raise_minimum_delegation_to_100m_lamports::id());
+        feature_set
+    }
+
+    /// The "old" behavior is before the stake minimum delegation was raised to either 0.1 or 1 SOL
+    fn feature_set_alternative_behavior() -> FeatureSet {
+        let mut feature_set = feature_set_new_behavior();
+        feature_set.deactivate(&feature_set::stake_raise_minimum_delegation_to_1_sol::id());
+        feature_set
+    }
```
[the remark this answers is not in the document]
Comment (michaelh-laine) [reply programs/stake/src/stake_instruction.rs:544]: Yes it seems a bit clunky to have this many tests repeated so many times. My main concern was ensuring all possible conditions in the `fn get_minium_delegation()` are tested but adding a direct test for this makes more sense then as you suggest. Will do that

**Comment by t-nelson** [comment]
where is the discussion justifying the new number?

File: programs/stake/src/lib.rs [hunk 2]
```diff
@@ -22,11 +22,17 @@ pub fn add_genesis_accounts(genesis_config: &mut GenesisConfig) -> u64 {
 /// The minimum stake amount that can be delegated, in lamports.
 /// NOTE: This is also used to calculate the minimum balance of a stake account, which is the
 /// rent exempt reserve _plus_ the minimum stake delegation.
+/// Multiple proposals to raise the delegation exist, enforce the feature with the highest minimum
```
[the remark this answers is not in the document]
Comment (michaelh-laine) [reply programs/stake/src/lib.rs:25]: I've rewritten the logic this way, but it's quite a big conditional, which might be undesirable. Would also appreciate a double-check on the logic, I don't think test can cover the "Activated slot"?

- if only the 1 SOL feature is active OR both are active AND 1 SOL feature activated slot is GREATER than 0.1 SOL feature activated slot (i.e. it's the more recently approved proposal) then return the 1 SOL min delegation
- else if the 0.1 SOL feature is active return 0.1 SOL (I think all the conditions in the first check should cover that you only need to check whether the feature is active here, we already know that the other feature is either not active or not more recent, if both are active)

File: programs/stake/src/stake_instruction.rs [hunk 3]
```diff
@@ -529,10 +529,25 @@ mod tests {
...
+
+    /// The "old" behavior is before the stake minimum delegation was raised to either 0.1 or 1 SOL
```
[the remark this answers is not in the document]
Comment (michaelh-laine) [reply programs/stake/src/stake_instruction.rs]: Reworked that, had forgotten to amend the copy pasted comments, but have now removed the "alternative" functions as moved the test to lib.rs

File: programs/stake/src/stake_instruction.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelh-laine) [reply programs/stake/src/stake_instruction.rs]: Moved the test to lib.rs, it's my first time doing anything with tests, I could compile the binaries locally with `./cargo build` but `./cargo test` returned some errors relating to `cc linkage` (so unrelated to my test I think). 

CI seems to have stalled? So not sure if the test is working

**Comment by michaelh-laine** [comment]
> where is the discussion justifying the new number?

https://github.com/solana-labs/solana/issues/22559#issuecomment-1223539906

and #vote-1-sol-min-delegations on Discord

0.1 SOL, based on current accounts, cuts out 60% of the accounts that would be affected by a 1 SOL minimum, while being just 10% of that number, so it strikes a balance of cutting unnecessary accounts while setting a threshold that will not be rendered unattainable as quickly as 1 SOL might be for some people in the future.

**Comment by michaelh-laine** [comment]
> > The CI is failing on an unrelated test relating to rpc pubsub?
> 
> Yeah... That test is notoriously flaky. I've retried the step, and it's going again. I'll handle babysitting the CI as needed. If it fails again and I haven't retried it already, feel free to tag/ping me here again.
> 
> Also, once you've updated the PR and its ready for another review, please click the "re-request review" so it'll show up in my GH tab for pending review requests. Thanks!

*ping* - CI timed out after over 2 hrs 🤷‍♂️

**Comment by t-nelson** [comment]
current accounts are already way too far into pathology. we need to over react with excessive back pressure, not stabilize the current state

File: programs/stake/src/lib.rs [hunk 4]
```diff
@@ -22,13 +22,69 @@ pub fn add_genesis_accounts(genesis_config: &mut GenesisConfig) -> u64 {
...
+                    &feature_set::stake_raise_minimum_delegation_to_100m_lamports::id(),
+                ))
+    {
         const MINIMUM_DELEGATION_SOL: u64 = 1;
         MINIMUM_DELEGATION_SOL * LAMPORTS_PER_SOL
+    } else if feature_set
+        .is_active(&feature_set::stake_raise_minimum_delegation_to_100m_lamports::id())
+    {
+        const MINIMUM_DELEGATION_LAMPORTS: u64 = 100_000_000;
+        MINIMUM_DELEGATION_LAMPORTS
```
[the remark this answers is not in the document]
Comment (michaelh-laine) [reply programs/stake/src/lib.rs:46]: Do we know the timing on mainnet as well though? Could the votes begin concurrently potentially? Or the first vote stalls/doesn't get approval before the second begins, but later crosses the line, causing it to be activated second?

**Comment by michaelh-laine** [comment]
Hey sorry for the lack of action here. My dev box melted down so I had to set things back up and in the mean time been giving this thought and don't think this PR makes sense right now. Pasting my comments from Discord here:

```
I'm probably gonna cancel the PR. 

Any stake minimum is a stop gap solution that doesn't address the underlying issue. If the cluster accepts 1 SOL I don't see the point of having a vote for 0.1. if 1 SOL is rejected we can consider a vote for 0.1.

However I'm far more interested in finding optimisations that make any concerns about the number of stake accounts of whatever size irrelevant
In the short term competing governance proposals for differing limits creates noise. I'm undecided on whether I'd vote in favour of 1 SOL minimum. I feel like we should be able to solve the performance issue before a 1 SOL minimum achieves anything, since it does nothing to reduce the current number of stake accounts. Long term a minimum of some sort makes sense as accounts with 1 lamport active stake never earn rewards due to rounding anyway, however any number (0.1, 1) will always feel random and a bit "wrong" but I have no solution for that
The main point for 0.1 SOL debate was that there was incorrect data indicating 1 SOL somehow encompasses 90% of stakes which is incorrect, and my argument is that 0.1 SOL does 60% of the job for 10% of the pain, but ultimately, meh, 1 SOL is a rounder figure and we already have that proposal in code so let's deal with that first
```

So I think as we already have a proposal passed on testnet and in the codebase we should see through the 1 SOL proposal, subject to which we could revisit an alternative amount. The current governance process doesn't lend itself suitably to multiple option voting. I think in future a discussion that involves the community and invites feedback before creating a proposal that is then voted upon would be preferable as an approach. That would lead to a much smoother voting process.

**Pull request closed without merging**